File: 0155-min-stack/0155-min-stack.java
class MinStack {

private Stack<Integer> minimums;
private Stack<Integer> stack;

/** initialize your data structure here. */
public MinStack() {
    minimums = new Stack<Integer>();
    stack = new Stack<Integer>();
}

public void push(int val) {
    // if we have a minimum and the new value is smaller than the current minimum
    if ( !minimums.isEmpty() && val <= minimums.peek() )
    {
        minimums.push(val);
    }
    // if this is the first value
    else if ( minimums.isEmpty() )
    {
        minimums.push(val);
    }
    stack.push(val);
}

public void pop() {
    // if what we're popping is the current minimum
    if ( (int)stack.peek() == (int)minimums.peek() )
    {
        minimums.pop();
    }
    stack.pop();
}

public int top() {
    return stack.peek();
}

public int getMin() {
    return minimums.peek(); 
	}
}
 /**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 0155-min-stack/0155-min-stack.py
class MinStack:
    def __init__(self):
        self.head = None

    def push(self, x):
        if self.head is None:
            self.head = self.Node(x, x, None)
        else:
            self.head = self.Node(x, min(x, self.head.min), self.head)

    def pop(self):
        self.head = self.head.next

    def top(self):
        return self.head.val

    def getMin(self):
        return self.head.min

    class Node:
        def __init__(self, val, min_val, next_node):
            self.val = val
            self.min = min_val
            self.next = next_node

        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()